File: LeetCodeProblemsGo/LeetCode840MagicSquaresInGrid/LeetCode840MagicSquaresInGrid.go
package LeetCode840MagicSquaresInGrid

func numMagicSquaresInside(grid [][]int) int {
	numsMap := make(map[int]any)
	heigth := len(grid)
	width := len(grid[0])
	result := 0
	for i := 1; i < heigth-1; i++ {
		for j := 1; j < width-1; j++ {
			if grid[i][j] != 5 {
				continue
			}
			numsMap = map[int]any{}
			isMagic := true
			for k := 0; k < 3; k++ {
				for l := 0; l < 3; l++ {
					if grid[i-1+k][j-1+l] > 9 {
						isMagic = false
						break
					}
					numsMap[grid[i-1+k][j-1+l]] = nil
				}
				if !isMagic {
					break
				}
			}
			if !isMagic {
				continue
			}
			if len(numsMap) != 9 {
				continue
			}
			if grid[i-1][j-1]+grid[i-1][j]+grid[i-1][j+1] != 15 {
				continue
			}
			if grid[i+1][j-1]+grid[i+1][j]+grid[i+1][j+1] != 15 {
				continue
			}
			if grid[i-1][j-1]+grid[i][j-1]+grid[i+1][j-1] != 15 {
				continue
			}
			if grid[i-1][j+1]+grid[i][j+1]+grid[i+1][j+1] != 15 {
				continue
			}
			if grid[i-1][j-1]+grid[i][j]+grid[i+1][j+1] != 15 {
				continue
			}
			if grid[i-1][j+1]+grid[i][j]+grid[i+1][j-1] != 15 {
				continue
			}

			result++
		}
	}
	return result
}

File: LeetCodeProblemsCMake/Problems/LeetCode840MagicSquaresInGrid.cpp
//
// Created by saeed on 12/30/25.
//

#include "LeetCode840MagicSquaresInGrid.h"

#include <unordered_map>
#include <unordered_set>

using namespace std;

int LeetCode840MagicSquaresInGrid::numMagicSquaresInside(vector<vector<int> > &grid) {
    unordered_set<int> nums{};
    int height = static_cast<int>(grid.size());
    int width = static_cast<int>(grid[0].size());
    int result = 0;
    bool isMagic = false;
    for (int i = 1; i < height-1; i++) {
        for (int j = 1; j < width-1; j++) {
            if (grid[i][j] != 5) {
                continue;
            }
            nums = {};
            isMagic = true;
            for (int k = 0; k < 3; ++k) {
                for (int l = 0; l < 3; ++l) {
                    if (grid[i -1 + k][j - 1 + l] > 9) {
                        isMagic = false;
                        break;
                    }
                    nums.insert(grid[i - 1 + k][j - 1 + l]);
                }
                if (!isMagic) {
                    break;
                }
            }
            if (!isMagic) {
                continue;
            }
            if (nums.size() != 9)
                continue;
            if (grid[i - 1][j - 1] + grid[i - 1][j] + grid[i - 1][j + 1] != 15) {
                continue;
            }
            if (grid[i + 1][j - 1] + grid[i + 1][j] + grid[i + 1][j + 1] != 15) {
                continue;
            }
            if (grid[i - 1][j - 1] + grid[i][j - 1] + grid[i + 1][j - 1] != 15) {
                continue;
            }
            if (grid[i - 1][j + 1] + grid[i][j + 1] + grid[i + 1][j + 1] != 15) {
                continue;
            }
            if (grid[i - 1][j - 1] + grid[i][j] + grid[i + 1][j + 1] != 15) {
                continue;
            }
            if (grid[i - 1][j + 1] + grid[i][j] + grid[i + 1][j - 1] != 15) {
                continue;
            }
            result++;
        }
    }
    return result;
}
